pre-commit/notebook_urls.py

```diff
@@ -0,0 +1,197 @@
+"""Simple check to ensure each code cell in a notebook is valid Python."""
+import argparse
+import json
+import re
+import subprocess
+import sys
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Set, Tuple
+
+
+def git_branch_name() -> str:
+    """Get the current branch name."""
+    return (
+        subprocess.check_output(["/usr/bin/git", "rev-parse", "--abbrev-ref", "HEAD"])
+        .decode()
+        .strip()
+    )
+
+
+def git_branch_modified_paths(from_ref: str, to_ref: str) -> Set[Path]:
+    """Get a set of file paths modified on this branch vs develop."""
+    from_to = f"{from_ref}...{to_ref}"
+    return {
+        Path(p)
+        for p in subprocess.check_output(
+            [
+                "/usr/bin/git",
+                "diff",
+                "--name-only",
+                from_to,
+            ]
+        )
+        .decode()
+        .strip()
+        .splitlines()
+    }
+
+
+def git_previous_commit_modified_paths() -> Set[Path]:
+    """Get a set of file paths modified in the previous commit."""
+    return {
+        Path(p)
+        for p in subprocess.check_output(
+            ["/usr/bin/git", "diff", "--name-only", "HEAD~"]
+        )
+        .decode()
+        .strip()
+        .splitlines()
+    }
+
+
+@dataclass(frozen=True)
+class PatternReplacement:
+    """Replacement dataclass.
+
+    Attributes:
+        pattern:
+            Regex pattern to match.
+        replacement:
+            Replacement string.
+        main_replacement:
+            Replacement string for main branch.
+
+    """
+
+    pattern: str
+    replacement: str
+    main_replacement: str = None
+
+
+MAIN_BRANCHES = ("master", "main")
+
+
+def main(files: List[Path], from_ref: str, to_ref: str) -> bool:
+    """Check that URLs in the notebook are relative to the current branch.
+
+    Args:
+        files:
+            List of files to check.
+        from_ref:
+            Reference to diff from.
+        to_ref:
+            Reference to diff to.
+
+    """
+    replacements = [
+        PatternReplacement(
+            pattern=(
+                r"(^\s*[!%]\s*)pip install "
```

Comment:
This only replaces `pip install ...` commands. Please can you also add replacements for links to GitHub and Colab at the start of each notebook. E.g., `https://github.com/TissueImageAnalytics/tiatoolbox/blob/develop/examples/01-wsi-reading.ipynb` should be replaced by `https://github.com/TissueImageAnalytics/tiatoolbox/blob/[dev-notebook-urls-hook](https://github.com/TissueImageAnalytics/tiatoolbox/tree/dev-notebook-urls-hook)/examples/01-wsi-reading.ipynb`. Same for Colab links.